File: ios/SpeedAlert/Location/SpeedFilter.swift
import Foundation

/// Turns noisy, intermittent GPS speed readings into a stable display value.
///
/// Two real-world problems this solves:
///
///  1. **Frozen value.** Core Location stops delivering fixes once you stop
///     moving (the distance filter), so the last speed — the one just before
///     you braked — stays on screen. A staleness watchdog decays it to 0 when
///     no fresh fix arrives.
///  2. **GPS noise floor.** Doppler speed rarely reads exactly 0 while
///     stationary; it hovers a few km/h. A dead-band snaps that to 0.
///
/// Rising speed is lightly smoothed to stop the digits flickering; falling
/// speed is *not* smoothed, so braking feels immediate.
struct SpeedFilter {
    /// Below this the vehicle is treated as stopped (GPS noise floor), km/h.
    var deadbandKph: Double = 3.0
    /// Drop to 0 if no fresh fix arrives for this long (seconds).
    var staleAfter: TimeInterval = 3.0

    private var kph: Double = 0
    private var lastFixAt: Date?

    /// Feed a raw fix. `speedMps` is `CLLocation.speed` (negative == invalid).
    mutating func update(speedMps: Double, at now: Date) {
        guard speedMps >= 0 else { return }   // invalid reading: keep last, watchdog handles it
        lastFixAt = now
        let raw = speedMps * 3.6
        if raw < deadbandKph {
            kph = 0                            // dead-band: snap the noise floor to zero
        } else if kph == 0 {
            kph = raw                          // starting to move: take it as-is
        } else {
            kph = kph * 0.6 + raw * 0.4        // light smoothing while cruising
        }
    }

    /// Value to show right now. Call this from a 1 s timer too, so a stopped
    /// car reads 0 even after the OS stops sending fixes.
    mutating func value(at now: Date) -> Double {
        if let last = lastFixAt, now.timeIntervalSince(last) > staleAfter {
            kph = 0
        }
        return kph
    }

    mutating func reset() {
        kph = 0
        lastFixAt = nil
    }
}

File: android/app/src/main/java/tw/speedalert/location/SpeedFilter.kt
package tw.speedalert.location

/**
 * Turns noisy, intermittent GPS speed readings into a stable display value.
 * Mirrors iOS `SpeedFilter`.
 *
 *  1. Frozen value: the OS stops sending fixes once the car stops moving, so
 *     the last speed stays on screen. The staleness watchdog decays it to 0.
 *  2. GPS noise floor: Doppler speed hovers a few km/h while stationary; the
 *     dead-band snaps that to 0.
 *
 * Rising speed is lightly smoothed; falling speed is not, so braking feels
 * immediate.
 */
class SpeedFilter(
    private val deadbandKph: Double = 3.0,
    private val staleAfterMs: Long = 3_000L,
) {
    private var kph: Double = 0.0
    private var lastFixAt: Long = 0L

    /** Feed a raw fix. `speedMps` is `Location.getSpeed()` (metres/second). */
    fun update(speedMps: Double, now: Long) {
        if (speedMps < 0) return            // invalid reading: keep last, watchdog handles it
        lastFixAt = now
        val raw = speedMps * 3.6
        kph = when {
            raw < deadbandKph -> 0.0         // dead-band: snap the noise floor to zero
            kph == 0.0 -> raw                // starting to move: take it as-is
            else -> kph * 0.6 + raw * 0.4    // light smoothing while cruising
        }
    }

    /** Value to show right now. Also call from a 1 s ticker so a stopped car reads 0. */
    fun value(now: Long): Double {
        if (lastFixAt != 0L && now - lastFixAt > staleAfterMs) kph = 0.0
        return kph
    }

    fun reset() {
        kph = 0.0
        lastFixAt = 0L
    }
}
